### PR TITLE
detach DragInteraction on DragBoxLayer.destroy()

### LINKED ISSUE #2553: .destroy doesn't .destroy dragBoxLayers
http://jsfiddle.net/fee6p9w0/6/
after clicking the destroy button, you can still drag and call the callback, even though the box doesn't show up any more.

### PR DESCRIPTION
**Pull request by ztsai**
closes #2553 
detach all dragInteractions on a DragBoxLayer when it's destroyed or detached

destroyed:
original: http://jsfiddle.net/fee6p9w0/6/
fixed: http://jsfiddle.net/fee6p9w0/16/
note that after clicking destroy button, there's still console log indicating interaction callback being fired in the original fiddle when drag events happen.

detached:
original: http://jsfiddle.net/fee6p9w0/24/
fixed: http://jsfiddle.net/fee6p9w0/25/
after clicking `detach` button there shouldn't be any console log while dragging, clicking `attach` will bring interaction and dbl back.

### CODE REVIEW

**Comment by aicioara** [comment]
Short description of what has been done is always welcome :)

File: src/components/dragBoxLayer.ts [hunk 1]
```diff
@@ -416,6 +416,12 @@ export module Components {
       this._setMovableClass();
       return this;
     }
+
+    public destroy() {
```
Comment (aicioara) [inline src/components/dragBoxLayer.ts]: Can we quickly check that this bug does not affect the `guideLineLayer` as well?

File: src/components/dragBoxLayer.ts [hunk 2]
```diff
@@ -416,6 +416,12 @@ export module Components {
...
+      super.destroy();
+      this._dragInteraction.detachFrom(this);
+    }
+
```
Comment (aicioara) [inline src/components/dragBoxLayer.ts]: Let's do exactly what the `dragLineLayer` is doing for completeness and consistency.

Comment (aicioara) [inline src/components/dragBoxLayer.ts]:
```typescript
        this._dragInteraction.offDragStart(interactionDragStartCallback);
        this._dragInteraction.offDrag(interactionDragCallback);
        this._dragInteraction.offDragEnd(interactionDragEndCallback);
        this._dragInteraction.detachFrom(this);

        this._dragStartCallbacks.forEach((callback) => this._dragStartCallbacks.delete(callback));
        this._dragCallbacks.forEach((callback) => this._dragCallbacks.delete(callback));
        this._dragEndCallbacks.forEach((callback) => this._dragEndCallbacks.delete(callback));
```

File: test/components/dragBoxLayerTests.ts [hunk 3]
```diff
@@ -106,6 +106,26 @@ describe("Interactive Components", () => {
...
+        let group = new Plottable.Components.Group([dbl]).renderTo(svg);
+        let target = group.background();
+        let onDragStartCallbackCalled = false;
+        let onDragCallbackCalled = false;
+        let onDragEndcallbackCalled = false;
+        dbl.onDragStart(() => onDragStartCallbackCalled = true);
+        dbl.onDrag(() => onDragCallbackCalled = true);
+        dbl.onDragEnd(() => onDragEndcallbackCalled = true);
+
+        dbl.destroy();
```
Comment (aicioara) [inline test/components/dragBoxLayerTests.ts]: assert that the callbacks were initially called

File: test/components/dragBoxLayerTests.ts [hunk 4]
```diff
@@ -106,6 +106,26 @@ describe("Interactive Components", () => {
         svg.remove();
       });
 
+      it("does not call callbacks after destroy() is called", () => {
+        let group = new Plottable.Components.Group([dbl]).renderTo(svg);
```
Comment (aicioara) [inline test/components/dragBoxLayerTests.ts]: why do we need the group?

Comment (ztsai) [inline test/components/dragBoxLayerTests.ts]: so that we can fake the mouse event somewhere (dbl.content() will be removed after destroy())

Comment (aicioara) [inline test/components/dragBoxLayerTests.ts]: Can we add a comment which explains this? It was not clear to me why is the group needed

**Comment by aicioara** [comment]
Apart from that it looks good.

**Comment by aicioara** [comment]
Oh, the fiddle is gone :(

**Comment by ztsai** [comment]
added comment and fixed fiddle

**Comment by crmorford** [comment]
Test Destroy() for Dragging/Resizing/Moving on DragBox/XDragBox/YDragBox
x: http://jsfiddle.net/fee6p9w0/17/
y: http://jsfiddle.net/fee6p9w0/18/
normal: http://jsfiddle.net/fee6p9w0/19/

Can we fix this for Detach(), too? Let me know if this is actually out of scope. http://jsfiddle.net/fee6p9w0/20/

**Comment by ztsai** [comment]
it's probably outside of the scope of the original issue, but I think it will make sense to fix it together
it seems like Detach() doesn't work on YDragBox as well

**Comment by ztsai** [comment]
QE notes above

**Comment by aicioara** [comment]
Apart from that and the bounds question it looks good.

File: src/components/dragBoxLayer.ts [hunk 5]
```diff
@@ -416,6 +428,32 @@ export module Components {
...
+      this._dragInteraction.detachFrom(this);
+      return super.detach();
+    }
+
+    public anchor(selection: d3.Selection<void>) {
+      this._dragInteraction.attachTo(this);
+      return super.anchor(selection);
+    }
+
+    private _resetBound() {
```
Comment (aicioara) [inline src/components/dragBoxLayer.ts]: Can we call it `_resetState()`, as ideally, we can make more resets here. 

By doing so, we make it more obvious that reseting the bounds is done for this purpose (resetting the state). Otherwise setting both bounds to 0 and the method `_resetBound()` are synonyms, so we don't get any benefit.
